analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/ControllersHaveTooManyResponsibilities.cs

```diff
@@ -0,0 +1,171 @@
+                    {
+                        CheckApiController(symbolStartContext, symbol);
+                    }
+                }, SymbolKind.NamedType);
+            }
+        });
+
+    private static void CheckApiController(SonarSymbolStartAnalysisContext symbolStartContext, INamedTypeSymbol controllerSymbol)
+    {
+        var memberNames = RelevantMemberNames(controllerSymbol);
+
+        if (memberNames.Count < 2)
+        {
+            return;
+        }
+
+        var dependencies = new ConcurrentStack<Dependency>();
+
+        symbolStartContext.RegisterCodeBlockStartAction<SyntaxKind>(codeBlockStartContext =>
+        {
+            if (BlockName(codeBlockStartContext.CodeBlock) is { } blockName)
+            {
+                codeBlockStartContext.RegisterNodeAction(c =>
+                {
+                    if (c.Node.GetName() is { } dependencyName && memberNames.Contains(dependencyName))
+                    {
+                        dependencies.Push(new(blockName, dependencyName));
+                    }
+                }, SyntaxKind.IdentifierName);
+            }
+        });
+
+        symbolStartContext.RegisterSymbolEndAction(symbolEndContext =>
+        {
+            var dependencySets = new DisjointSets(memberNames);
+            foreach (var dependency in dependencies)
+            {
+                dependencySets.Union(dependency.From, dependency.To);
+            }
+
+            var responsibilityGroups = dependencySets.GetAllSets();
+            if (responsibilityGroups.Count > 1)
+            {
+                var secondaryLocations = SecondaryLocations(controllerSymbol, responsibilityGroups);
+                foreach (var primaryLocation in LocationIdentifiers<ClassDeclarationSyntax>(controllerSymbol))
+                {
+                    symbolEndContext.ReportIssue(Diagnostic.Create(
+                        Rule,
+                        primaryLocation,
+                        secondaryLocations.ToAdditionalLocations(),
+                        secondaryLocations.ToProperties(),
+                        responsibilityGroups.Count));
+                }
+            }
+        });
+    }
+
+    private static string BlockName(SyntaxNode block) =>
+        block switch
+        {
+            MethodDeclarationSyntax method => method.GetName(),
+            AccessorDeclarationSyntax { Parent.Parent: PropertyDeclarationSyntax property } => property.GetName(),
+            _ => null
+        };
+
+    private static IEnumerable<Location> LocationIdentifiers<T>(ISymbol symbol) where T : SyntaxNode =>
+        symbol.DeclaringSyntaxReferences.Select(x => x.GetSyntax()).OfType<T>().Select(x => x.GetIdentifier()?.GetLocation());
+
+    private static ImmutableHashSet<string> RelevantMemberNames(INamedTypeSymbol symbol)
+    {
+        var builder = ImmutableHashSet.CreateBuilder<string>();
+        foreach (var member in symbol.GetMembers())
```

Comment:
This is a tad bit expensive.
Maybe it's better to extract it to a field and set it once in the initialization?

Reply:
What do you mean by this?
`GetMembers()` is called once per `symbol`.

Reply:
Ah that's a fair point, I see what you mean.
I agree with your suggestion, with one caveat:

It cannot be a field, as it should be scoped to the registration callback. 
The analyzer should not have any state.
We could do `var members = symbol.GetMembers();` at the top and pass it to both methods you mentioned.

Reply:
Indeed! Thanks @gregory-paidis-sonarsource .

Reply:
Roslyn has a [specific API](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.inamespaceortypesymbol.getmembers?view=roslyn-dotnet-4.7.0#microsoft-codeanalysis-inamespaceortypesymbol-getmembers(system-string)) for retrieval by name, that works with overloads.
I have used that one instead.